man/man1/ccdstat.1
.TH CCDSTAT 1NEMO "1 December 2022"

.SH "NAME"
ccdstat \- statistics (1st through 4th moment) and chi2

.SH "SYNOPSIS"
.PP
\fBccdstat in=\fPimage [parameter=value]

.SH "DESCRIPTION"
\fBccdstat\fP calculates all moments (up to fourth order) of the values
in an image/cube, 
and calculate some human understandable numbers (mean, dispersion, skewness,
curtosis) from them. It can optionally use a weighting map, in case the 
map needs to be weighted per pixel.
.PP
Optionally, if the map contains an error normalized difference between an
observation and model, it can also compute a chi2 goodness of
fit if the number of 
of fitting parameters is given
(\fBnpar\fP, must be > 0 to trigger this option),
as well as the number of independant points per beam (\fBnppb\fP):
.nf
    chi2 = sum_{i,j}(map(i,j)^2)/nppb/(N/nppb-npar)
.fi
where N is the number of non-bad values in the map. The input 
map is now assumed to contain a dimensionless \fI(obs-mod)/sigma\fP

.SH "PARAMETERS"
.so man1/parameters
.TP 20
\fBin=\fIin-file\fP
input file, in \fIImage(5)\fP format [no default]
.TP
\fBmin=\fImin_value\fP
If given, all values below this value are ignored in the statistics. 
By default no lower limit is used.
.TP
\fBmax=\fP\fImax_value\fP
If given, all values above this value are ignored in the statistics.
By default no upper limit is used.
.TP
\fBbad=\fIbad_value\fP
If given, all values identical to this value are ignored in the statistics.
Note that, apart for bad=0, it is dangerous to use equating to a real value
as a masking operating.
.TP
\fBwin=\fIweight-file\fP
optional input file, in \fIImage(5)\fP format, of which the mapvalues
are used as a relative weight factor. Should be all positive, with 0
meaning no weight.
[no default]
.TP
\fBnpar=\fP
Number of fitting parameters. If given > 0, this will trigger the
computation of chi2 goodness of fit, instead of the normal statistics.
.TP
\fBnppb=\fP
Number of (independant) points per beam. Only used in the computation
for chi2, and 
.TP
\fBmedian=t|f\fP
Optional display of the median value
.TP
\fBtorben=t|f\fP
Use the \fItorben\fP method to compute the median. [Default: f]
.TP
\fBrobust=t|f\fP
Conmpute robust mean, median and dispersion. [Default: f]
.TP
\fBsratio=t|f\fP
Conmpute signed flux ratio (FP-FN)/(FP+FN).. [Default: f]
.TP
\fBmmcount=t|f\fP
Count occurances of min and max
.TP
\fBmaxpos=t|f\fP
Report the location where the max occured. [Default: f]
.TP
\fBhalf=f|t\fP
Only use half (negative) values and symmetrize them 
.TP
\fBmaxmom=\fP
Maximum moments it should compute. The current maximum is 4, moments up
through the kurtosis are computed. For maxmom<0 no work is done, except
for reading the file, which can be used to benchmark your disk I/O for
large NEMO images.  [Default: 4].
.TP
\fBignore=t|f\fP
This will ignore the cell width when summing emission for any axis
that has only one cell (assumed infinity, or a result of \fIccdmom(1NEMO)\fP).
Setting this to false, will not ignore that cell width. [Default : t]
.TP
\fBsort=t|f\fP
Sorting routine (not activated yet) 
Flux not implemented yet. [
Default: f]
.TP
\fBplanes=-1|0|start:end:step\fP
Either the whole cube is taken as one (-1, the default),
or planes can be selected.  0 means all planes, if a
start:end:step is selected, only those planes are. Step defaults to 1.
Default: -1
.TP
\fBtab=\fP
If a table output filename is given, data values are written here. otherwise
on screen.
.TP
\fBqac=t|f\fP
If selected, a QAC listing is selected. This gives the mean, rms, min, max, etc.
See \fIqac_stats(5NEMO)\fP for more info.
[Default: f]
.TP
\fBlabel=\fP
Descriptive label for QAC output. Normally the filename is used, but for awkward long
or short (a pipe) filenames this can be useful. Default: \fIin_file\fP

.SH "EXAMPLE"
.nf
  % ccdstat image.ccd min=-2
Min=-1.879409  Max=2.265516
Number of points     : 95
Mean and dispersion  : 0.097891 0.956030
Skewness and kurtosis: -0.059936 -0.654776
5/100 out-of-range points discarded
.fi
Here is an example how to created a weight file which is a circular annulus:
.nf
  % ccdmath out=map1 "fie=iflt(%r,50,1,0)" size=200,200 crpix=100,100
  % ccdmath out=map2 "fie=iflt(%r,20,1,0)" size=200,200 crpix=100,100
  % ccdmath map1,map2 map3 %1-%2
.fi
Here is an two masking example: firstignoring the exact value of 0, 
followed by an example of a small range around 0:
.nf
  % ccdstat map1 bad=0
  % ccdmath map1 - 'iflt(abs(%1),0.001,0,%1)' | ccdstat - bad=0
.fi
Here is an example of calculating the chi2 goodness of fit for
a gaussian distribution:
.nf
  % ccdmath "" - "rang(0,1)" | ccdstat - npar=1	
Generating a map from scratch
chi2= 1.03431
df= 98
.fi

.SH "SEE ALSO"
snapccd(1NEMO), image(5NEMO), qac_stats(5NEMO)

.SH "AUTHOR"
Peter Teuben

.SH "FILES"
.nf
.ta +2.5i
src/image/misc	ccdstat.c
.fi

.SH "UPDATE HISTORY"
.nf
.ta +1.5i +5.5i
 4-Oct-88	V1.0: Created		PJT
9-may-03	V1.3: added bad=	PJT
5-jun-03	V1.4: added win=	PJT
14-nov-04	V1.5: added the chi2 trigger with npar= and nppb=	PJT
24-may-06	V1.8: added mmcount	PJT
15-oct-11	V1.10: added maxmom= and bench options	PJT
14-feb-13	V2.0:  ignore=t to properly handle units	PJT
4-dec-2020	V3.8: added qac=	PJT
1-dec-2022	V3.12: added sratio=	PJT
.fi


src/image/misc/ccdstat.c
/* 
 * CCDSTAT
 *
 *	 4-oct-88  V1.0 created
 *	 5-nov-93  V1.1 nemo V2, and using moment.h
 *      16-sep-95  V1.2 added min/max
 *       9-may-03  V1.3 added bad=, and added #points,
 *       5-jun-03  V1.4 added win=, a weight map
 *      14-nov-04   1.5 provide nppb= correction for chi2 computation   PJT
 *       5-jan-05   1.6 added a total
 *      30-jan-05   1.7 added an optional median
 *      24-may-06   1.8 add mmcount=
 *       6-nov-06   1.9 output the total mass/luminosity as well as sum of densities
 *      13-feb-13   2.0 ignore axis if N=1 
 *      17-apr-13   3.0 allow each Z plane to be treated seperately
 *      23-apr-13       half= option, use compute_robust_moment()
 *       6-jun-13   3.2 added torben, cleanup a bit
 *      28-feb-14   3.3 added tab=
 *      25-aug-14   3.5 added maxpos=, fixed duplicate header, so format='ascii.commented_header'
 *                      works in astropy
 *    26-dec-2019   3.6 (not finished yet) enable some openmp sections of code
 *    11-oct-2020   3.7 optimized memory usage, speed up median computation
 *     4-dec-2020   3.8 qac mode
 *     1-dec-2022   3.12 qac mode when planes >= 0
 */
 
#include <stdinc.h>
#include <getparam.h>
#include <filestruct.h>
#include <image.h>
#include <moment.h> 

string defv[] = {
    "in=???\n       Input image filename",
    "min=\n         Minimum overrride",
    "max=\n         Maximum overrride",
    "bad=\n         Use this as masking value to ignore data",
    "win=\n         Optional input map for weights",
    "npar=0\n       Number of fitting parameters assumed for chi2 calc (full data only)",
    "nppb=1\n       Optional correction 'number of points per beam' for chi2 calc",
    "median=f\n     Optional display of the median value",
    "torben=f\n     Use torben method for median instead",
    "robust=f\n     Compute robust median",
    "sratio=f\n     Optional display of the signed fluxes (FP-FN)/(FP+FN) ratio",
    "mmcount=f\n    Count occurances of min and max",
    "maxpos=f\n     Add location of where the max occured",
    "half=f\n       Only use half (negative) values and symmetrize them",
    "maxmom=4\n     Control how many moments are computed",
    "ignore=t\n     (for summing) Ignore cell width when N=1 (assumed infinity)",
    "sort=qsort\n   Sorting routine (not activated yet)",
    "planes=-1\n    -1: whole cube in one      0=all planes   start:end:step = selected planes",
    "tab=\n         If given, print out data values",
    "qac=f\n        QAC mode listing mean,rms,min,max",
    "label=\n       QAC label",
    "VERSION=3.12\n 1-dec-2022 PJT",
    NULL,
};

string usage="basic statistics of an image, optional chi2 calculation";

string	infile;	        		/* file names */
stream  instr;				/* file streams */
stream  tabstr = NULL;                  /* output table? */

imageptr iptr=NULL;			/* will be allocated dynamically */
imageptr wptr=NULL;                     /* optional weight map */

int    nx,ny,nz,nsize;			/* actual size of map */
double xmin,ymin,zmin,dx,dy,dz;
double size;				/* size of frame (square) */
double cell;				/* cell or pixel size (square) */
int *planes = NULL;                     /* selected planes , if used */

real get_median(int n, real *x);        /* uses an in place sort */

/* median.c */
extern real smedian(int,real*);
extern real smedian_q1(int,real*);
extern real smedian_q3(int,real*);
extern real median_torben(int n, real *x, real xmin, real xmax);

void nemo_main(void)
{
    int  i, j, k, ki;
    real x, y, z, xmin, xmax, mean, sigma, skew, kurt,  bad, w, *data;
    real dmin, dmax;
    real sum, sov, q1, q2, q3, tm;
    Moment m;
    bool Qmin, Qmax, Qbad, Qw, Qmedian, Qrobust, Qtorben, Qmmcount, Qsratio;
    bool Qx, Qy, Qz, Qone, Qall, Qign = getbparam("ignore");
    bool Qhalf = getbparam("half");
    bool Qmaxpos = getbparam("maxpos");
    bool Qac = getbparam("qac");
    string qac_label;
    real nu, nppb0, nppb = getdparam("nppb");
    int npar = getiparam("npar");
    int ngood;
    int ndat = 0;
    int nplanes;
    int min_count, max_count;
    int maxmom = getiparam("maxmom");
    int maxpos[2];
    char slabel[32];


    instr = stropen (getparam("in"), "r");
    read_image (instr,&iptr);
    strclose(instr);
    nx = Nx(iptr);	
    ny = Ny(iptr);
    nz = Nz(iptr);
    if (Beamx(iptr)>0 && Beamy(iptr)>0) {
      // 1.13309 = pi/(4*ln(2))
      nppb0 = 1.13309 * Beamx(iptr) * Beamy(iptr) / (Dx(iptr)*Dy(iptr));
      nppb0 = ABS(nppb0);
      dprintf(1,"nppb = %g\n",nppb0);
    } else
      nppb0 = 1.0;
    
    dprintf(1,"# data order debug:  %f %f\n",Frame(iptr)[0], Frame(iptr)[1]);
    if (hasvalue("tab")) tabstr = stropen(getparam("tab"),"w");

    planes = (int *) allocate((nz+1)*sizeof(int));
    nplanes = nemoinpi(getparam("planes"),planes,nz+1);
    Qall = (planes[0]==-1);
    if (planes[0]==0) {
      Qone = FALSE;
      nplanes = nz;
      for (k=0; k<nz; k++)
	planes[k] = k;
    } else if (!Qall) {
      Qone = (!Qall && nplanes==1);
      for (k=0; k<nplanes; k++) {
	if (planes[k]<1 || planes[k]>nz) error("%d is an illegal plane [1..%d]",planes[k],nz);
	planes[k]--;
      }
    }
 
    if (hasvalue("win")) {
      instr = stropen (getparam("win"), "r");
      read_image (instr,&wptr);
      strclose(instr);
      if (Nx(iptr) != Nx(wptr)) error("X sizes of in=/win= don't match");
      if (Ny(iptr) != Ny(wptr)) error("Y sizes of in=/win= don't match");
      if (Nz(iptr) != Nz(wptr)) error("Z sizes of in=/win= don't match");
      Qw = TRUE;
    } else
      Qw = FALSE;

    if (hasvalue("label"))
      qac_label = getparam("label");
    else
      qac_label = getparam("in");

    Qmin = hasvalue("min");
    if (Qmin) xmin = getdparam("min");
    Qmax = hasvalue("max");
    if (Qmax) xmax = getdparam("max");
    Qbad = hasvalue("bad");
    if (Qbad) bad = getdparam("bad");
    Qsratio = getbparam("sratio");
    Qmedian = getbparam("median");
    Qrobust = getbparam("robust");
    Qtorben = getbparam("torben");
    Qmmcount = getbparam("mmcount");    
    if (Qtorben) Qmedian = TRUE;
    if (Qmedian || Qrobust || Qtorben) {
      ndat = nx*ny*nz;
      dprintf(1,"Need spare array size %d\n",ndat);
      data = (real *) allocate(ndat*sizeof(real));
    }

    sov = 1.0;       /* volume of a pixel/voxel */
    Qx = Qign && Nx(iptr)==1;
    Qy = Qign && Ny(iptr)==1;
    Qz = Qign && Nz(iptr)==1;
    sov *= Qx ? 1.0 : Dx(iptr);
    sov *= Qy ? 1.0 : Dy(iptr);
    sov *= Qz ? 1.0 : Dz(iptr);
    strcpy(slabel,"*");
    if (!Qx) strcat(slabel,"Dx*");
    if (!Qy) strcat(slabel,"Dy*");
    if (!Qz) strcat(slabel,"Dz*");
    /* make it always 10 in lenght */
    for (i=strlen(slabel); i<10; i++)
      strcat(slabel," ");
    
    if (maxmom < 0) {
      warning("No work done, maxmom<0");
      stop(0);
    }

    if (Qall) {                 /* treat cube as one data block */

      ini_moment(&m,maxmom,ndat);
      ngood = 0;
#if 0
      // not working yet
      #pragma omp parallel \
	shared(ngood, nx,ny,nz, iptr,wptr,m, xmin,xmax,bad,Qhalf,Qmin,Qmax,Qbad,Qw,data,tabstr,Qmedian) \
        private(i,j,k,x,w)
      #pragma omp for
#endif      
      for (k=0; k<nz; k++) {
	for (j=0; j<ny; j++) {
	  for (i=0; i<nx; i++) {
            x =  CubeValue(iptr,i,j,k);    // iptr->cube[k,j,i]
	    if (isnan(x)) continue;
	    if (Qhalf && x>=0.0) continue;
            if (Qmin  && x<xmin) continue;
            if (Qmax  && x>xmax) continue;
            if (Qbad  && x==bad) continue;
	    w = Qw ? CubeValue(wptr,i,j,k) : 1.0;
            accum_moment(&m,x,w);
	    if (Qhalf && x<0) accum_moment(&m,-x,w);
	    if (Qmedian) data[ngood++] = x;
	    if (tabstr) fprintf(tabstr,"%g\n",x);
	  }
	}
      }
      
      if (npar > 0) {
	nu = n_moment(&m)/nppb - npar;
	if (nu < 1) error("%g: No degrees of freedom",nu);
	printf("chi2= %g\n", show_moment(&m,2)/nu/nppb);
	printf("df= %g\n", nu);
      } else {
	nsize = nx * ny * nz;
	sum = mean = sigma = skew = kurt = 0;
	if (maxmom > 0) {
	  mean = mean_moment(&m);
	  sum = show_moment(&m,1);
	}
	if (maxmom > 1)
	  sigma = sigma_moment(&m);
	if (maxmom > 2)
	  skew = skewness_moment(&m);
	if (maxmom > 3)
	  kurt = kurtosis_moment(&m);

	if (Qac) {
	  if (Qrobust) {
	    compute_robust_moment(&m);
	    printf("QAC_STATS: %s %g %g %g %g  %g %g  %d\n",
		   qac_label, mean_robust_moment(&m), sigma_robust_moment(&m), min_moment(&m), max_moment(&m),
		   sum_moment(&m), sratio_moment(&m), n_robust_moment(&m));
	  } else
	    printf("QAC_STATS: %s %g %g %g %g  %g %g  %d\n",
		   qac_label, mean, sigma, min_moment(&m), max_moment(&m),
		   sum_moment(&m), sratio_moment(&m) ,n_moment(&m));

	  return;
	}
	
	printf ("Number of points       : %d\n",n_moment(&m));
	printf ("Min and Max            : %f %f\n",min_moment(&m), max_moment(&m));
	printf ("Mean and dispersion    : %f %f\n",mean,sigma);
	printf ("Skewness and kurtosis  : %f %f\n",skew,kurt);
	printf ("Sum and Sum%s  : %f %f\n",slabel,sum,sum*sov);   /* align trick */
	printf ("Points per beam (map)  : %g\n",nppb0);
	if (Qmedian) {
	  if (Qtorben) {
	    printf ("Median Torben          : %f (%d)\n",median_torben(ngood,data,min_moment(&m),max_moment(&m)),ngood);
	  } else {
	    printf ("Median                 : %f\n",get_median(ngood,data));  // in-place sort
	    q2 = smedian(ngood,data);                                         // slow if not in-place sorted before
	    q1 = smedian_q1(ngood,data);
	    q3 = smedian_q3(ngood,data);
	    tm = (q1 + 2*q2 + q3 ) / 4.0;
	    printf ("Q1,Q2,Q3               : %f %f %f\n",q1,q2,q3);
	    printf ("TriMean                : %f\n",tm);
	  }
#if 0
	  if (ndat>0)  // very slow
	    printf ("MedianL                : %f\n",median_moment(&m));       // slow
#endif
	}
	if (Qrobust) {
	  compute_robust_moment(&m);
	  printf ("N Robust               : %d\n",n_robust_moment(&m));
	  printf ("Mean Robust            : %f\n",mean_robust_moment(&m));
	  printf ("Sigma Robust           : %f\n",sigma_robust_moment(&m));
	  printf ("Median Robust          : %f\n",median_robust_moment(&m));
	}

	if (Qmmcount) {
	  min_count = max_count = 0;
	  xmin = min_moment(&m);
	  xmax = max_moment(&m);
	  for (i=0; i<nx; i++) {
	    for (j=0; j<ny; j++) {
	      for (k=0; k<nz; k++) {
		x =  CubeValue(iptr,i,j,k);
		if (x==xmin) min_count++;
		if (x==xmax) max_count++;
	      }
	    }
	  } /* i */
	  printf("Min_Max_count         : %d %d\n",min_count,max_count);
	}
	printf ("%d/%d out-of-range points discarded\n",nsize-n_moment(&m), nsize);
      }

      free_moment(&m);

    } else {             /* treat each plane seperately */

      /* tabular output, one line per (selected) plane */

      printf("# iz z min  max  N  mean sigma skew kurt sum sumsov");
      if (Qsratio) printf(" sratio");
      if (Qmedian) printf(" med1 med2");
      if (Qrobust) printf(" rN rmean rsigma rmed]");
      if (Qmaxpos) printf(" maxposx maxposy");
      printf("\n");

      ini_moment(&m,maxmom,ndat);
      for (ki=0; ki<nplanes; ki++) {
	reset_moment(&m);
	k = planes[ki];
	z = Zmin(iptr) + k*Dz(iptr);
	ngood = 0;
	for (j=0; j<ny; j++) {
	  for (i=0; i<nx; i++) {
            x =  CubeValue(iptr,i,j,k);
	    if (isnan(x)) continue;
            if (Qmin && x<xmin) continue;
            if (Qmax && x>xmax) continue;
            if (Qbad && x==bad) continue;
	    if (Qmaxpos) {
	      if (i==0 && j==0) { dmax = x; maxpos[0] = 0; maxpos[1] = 0;}
	      else if (x>dmax) {  dmax = x; maxpos[0] = i; maxpos[1] = j;}
	    }
	    w = Qw ? CubeValue(wptr,i,j,k) : 1.0;
            accum_moment(&m,x,w);
	    if (Qmedian) data[ngood++] = x;
	  }
	}

	nsize = nx * ny * nz;
	sum = mean = sigma = skew = kurt = 0;
	if (maxmom > 0) {
	  mean = mean_moment(&m);
	  sum = show_moment(&m,1);
	}
	if (maxmom > 1)
	  sigma = sigma_moment(&m);
	if (maxmom > 2)
	  skew = skewness_moment(&m);
	if (maxmom > 3)
	  kurt = kurtosis_moment(&m);
	if (n_moment(&m) == 0) {
	  printf("# %d no data\n",k+1);
	  continue;
	}
	printf("%d %f  %f %f %d  %f %f %f %f  %f %f",
	       k+1, z, min_moment(&m), max_moment(&m), n_moment(&m),
	       mean,sigma,skew,kurt,sum,sum*sov);
	if (Qsratio) {
	  printf ("   %f",sratio_moment(&m));
	}
	if (Qmedian) {
	  printf ("   %f",get_median(ngood,data));
	  if (ndat>0) printf (" %f",median_moment(&m));
	}
	if (Qrobust) {
	  compute_robust_moment(&m);
	  printf ("   %d %f %f %f",n_robust_moment(&m), mean_robust_moment(&m),
		  sigma_robust_moment(&m), median_robust_moment(&m));
	}
	if (Qmaxpos) {
	  printf("   %d %d",maxpos[0],maxpos[1]);
	}
#if 0	  
	if (Qmmcount) {
	  min_count = max_count = 0;
	  xmin = min_moment(&m);
	  xmax = max_moment(&m);
	  for (i=0; i<nx; i++) {
	    for (j=0; j<ny; j++) {
	      x =  CubeValue(iptr,i,j,k);
	      if (x==xmin) min_count++;
	      if (x==xmax) max_count++;
	    }
	  } /* i,j */
	  printf(" %d %d",min_count,max_count);
	}
	printf ("%d/%d out-of-range points discarded\n",nsize-n_moment(&m), nsize);
#endif
	printf("\n");
      } /* ki */
      free_moment(&m);
    }
}


/*   @todo:   this could be placed in median.c?
 *   local version of a median. The one in moment.c and median.c are sorting
 *   pointers , this one sorts the data 
 */

int compar_real(const void *va, const void *vb)
{
  real *a = (real *) va;
  real *b = (real *) vb;
  return *a < *b ? -1 : *a > *b ? 1 : 0;
}

real get_median(int n, real *x)
{
  dprintf(1,"get_median: n=%d\n",n);

  qsort(x,n,sizeof(real),compar_real);
  if (n % 2)
    return  x[(n-1)/2];
  else
    return 0.5 * (x[n/2] + x[n/2-1]);
}
